### PR TITLE
Extend set_index for N-D Arrays

### PR DESCRIPTION
**Pull request by rkshthrmsh**
- Add suppport for N-dimensional indices.
- Add more information to assertion messages.

### CODE REVIEW

File: magma/primitives/set_index.py [hunk 1]
```diff
@@ -2,27 +2,42 @@
 from magma.bits import UInt
 from magma.bitutils import clog2
 from magma.primitives import mux
+from typing import Union, List
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:5]:
```suggestion
from collections.abc import Sequence
from typing import Union

from magma.bits import UInt
from magma.bitutils import clog2
from magma.primitives import mux
```

File: magma/primitives/set_index.py [hunk 2]
```diff
@@ -2,27 +2,42 @@
...
 
-
-def set_index(target: Array, value, idx: UInt):
+def set_index(target: Array, value, idx: Union[UInt, List[UInt]]):
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:7]:
```suggestion
def set_index(target: Array, value, idx: Union[UInt, Sequence[UInt]]):
```

File: magma/primitives/set_index.py [hunk 3]
```diff
@@ -2,27 +2,42 @@
...
     """
-    Returns a new value where index `idx` of value `target` is set to `value`
-    * `target` - a value of type `Array[N, T]`
+    Returns a new value where index `idx` of `target` is set to `value`
+    * `target` - a value of type `Array[N, T]` or `Array[(N, M, ..), T]`
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:10]:
```suggestion
    * `target` - a value of type `Array[N, T]` or `Array[(N, M, L, ...), T]`
```

File: magma/primitives/set_index.py [hunk 4]
```diff
@@ -2,27 +2,42 @@
...
     * `value` - a value of type `T`
-    * `idx` - a value of type `UInt[clog2(N)]`
+    * `idx` - a value of type `UInt[clog2(N)]` or `[UInt[clog2(M)], UInt[clog2(N)], ..]`
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:12]:
```suggestion
    * `idx` - a value of type `UInt[clog2(N)]` or `(UInt[clog2(L)], UInt[clog2(M), UInt[clog2(N)], ...)`
```

File: magma/primitives/set_index.py [hunk 5]
```diff
@@ -2,27 +2,42 @@
...
+    NOTE: Ordering of indices in `idx` is reverse of ordering in N-D Array definition.
+    More details: https://github.com/phanrahan/magma/issues/1310
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:14]:
```suggestion
    NOTE(rkshthrmsh): Ordering of indices in `idx` is reverse of ordering in N-D Array definition.
    For more details see: https://github.com/phanrahan/magma/issues/1310.
```

File: magma/primitives/set_index.py [hunk 6]
```diff
@@ -2,27 +2,42 @@
...
+        if len(idx) != clog2(len(target_T)):
+            raise TypeError(
+                (f"Expected number of `idx` ({len(idx)}) bits to map to the length of"
+                f"`target` ({clog2(len(target_T))})"))
+
+        return target_T([
+            mux([elem, value], idx == i) for i, elem in enumerate(target)
+        ])
+    else:
+        raise TypeError(f"Expected `idx` ({type(idx)}) to be UInt or List[UInt, ...] ")
```
Comment (rsetaluri) [inline magma/primitives/set_index.py:43]: Let's flip these

```suggestion
    if isinstance(idx, UInt):
        target_T = type(target)
        if not isinstance(value, target_T.T):
            raise TypeError(
                f"Expected `value` ({type(value)}) to be the same type as"
                f"`target`'s contents ({target_T.T})"
            )
        if not isinstance(idx, UInt):
            raise TypeError("Expected `idx` ({type(idx)}) to be a UInt")
        if len(idx) != clog2(len(target_T)):
            raise TypeError(
                f"Expected number of `idx` ({len(idx)}) bits to map to the length of"
                f"`target` ({clog2(len(target_T))})")
            )
        return target_T([
            mux([elem, value], idx == i) for i, elem in enumerate(target)
        ])
    if isinstance(idx, Sequence):
        if len(idx) == 1:
            return set_index(target, value, idx[0])
        return set_index(target, set_index(target[idx[0]], value, idx[1:]), idx[0])
    raise TypeError(f"Expected `idx` ({type(idx)}) to be UInt or List[UInt, ...] ")
```

**Review by rsetaluri — CHANGES_REQUESTED** [review]

**Comment by rkshthrmsh** [comment]
@rsetaluri, I have committed all the suggestions. Thanks!

**Comment by rkshthrmsh** [comment]
@rsetaluri, I have made the necessary changes to the test file. However, I'm unable to run it because `pip install fault` fails with:
```
 ERROR: Failed building wheel for coreir
  Building wheel for DeCiDa (pyproject.toml) ... done
  Created wheel for DeCiDa: filename=DeCiDa-1.1.5-py3-none-any.whl size=13947008 sha256=8dda2758409edef16994bd393ab64c3e7f6e25e80a597367ac0f1fb4a18f0595
  Stored in directory: /Users/rkshthrmsh/Library/Caches/pip/wheels/64/21/c3/f58e97f65bdac62c9ac1e98711dc44d30300e8801a4eccc2f8
Successfully built fault DeCiDa
Failed to build coreir
ERROR: Could not build wheels for coreir, which is required to install pyproject.toml-based projects
```

How can I get around this?

**Comment by rkshthrmsh** [comment]
@rsetaluri, added test for N-D `set_index()`

**Comment by rkshthrmsh** [comment]
@rsetaluri / @leonardt, any update on this PR?

**Review by leonardt — APPROVED** [review]

**Comment by leonardt** [comment]
@rkshthrmsh can you either merge or rebase on the latest master just to make sure it's working with the latest head?

**Comment by rkshthrmsh** [comment]
Closing this as it is merged in #1353

**Pull request closed without merging**